### ISSUE TITLE
Suggestion details layout issues

### ISSUE THREAD

**Issue by alexandrudima** [microsoft/monaco-editor#474]
* On Edge
* run `monaco-editor/test/index.html?editor=dev`
* Type something like `window.add`
* Press Ctrl+Space

![image](https://user-images.githubusercontent.com/5047891/27442861-b415db76-5771-11e7-8f97-8e0d859a6f26.png)

* Focus out or close suggestions. Press Ctrl+Space again:

![image](https://user-images.githubusercontent.com/5047891/27443185-905e2e80-5772-11e7-8989-1fea3b8aed75.png)

---

I could reproduce this on Chrome too.

The only difference that I can think of is the editor `fixedOverflowWidgets` option, which is `true` in VS Code and `false` in the standalone editor.

Here are the scenarios where the suggest widget should be positioned correctly:
* when the editor is in a fixed element: `mouse-fixed.html`
* when the editor is in a scrollable body: `mouse-scrollable-body.html`
* when the editor is in a scrollable element: `mouse-scrollable-element.html`

And the browser matrix is: IE11, Edge, latest Firefox, latest Chrome, Safari.

The setup details for running the editor from source are [here](https://github.com/Microsoft/monaco-editor/blob/master/CONTRIBUTING.md#running-the-editor-from-source). Remember to click on the `dev` to run the editor from sources!

**Comment by ramya-rao-a** [comment]
Fixed with https://github.com/Microsoft/vscode/commit/80a557774822998c698ac0f7eee44043de5f5a21

The issue was that I used the StorageService to store user choice of expanding docs.
Storage Service is not available in Monaco editor

**Comment by alexandrudima** [comment]
Thank you for looking into this promptly and thank you for fixing the details not showing up again.

The test pages can be edited to change the language to javascript and then JavaScript code can be copy-pasted in the editor to try out suggestion details. e.g. for `mouse-fixed`.
![image](https://user-images.githubusercontent.com/5047891/27468285-4962a502-57e9-11e7-904e-d6e259d6c745.png)




The following issues remain:

### 1. The suggestions are positioned to the bottom (away from the cursor)

| v0.8 on Chrome | master on Chrome | VS Code |
|---|---|---|
| ![image](https://user-images.githubusercontent.com/5047891/27468523-9aeb26e6-57ea-11e7-857f-a05342920c1c.png) | ![image](https://user-images.githubusercontent.com/5047891/27468184-a4728a3a-57e8-11e7-83ab-01e6c8277306.png) | ![image](https://user-images.githubusercontent.com/5047891/27468153-6e379ca8-57e8-11e7-83b2-01d1fcffd109.png) |


### 2. Once details are shown once, the suggestions permanently remain positioned wrongly
* test page `test/mouse-scrollable-element.html` with an edit to change the language from `xml` to `javascript`.

| v0.8 on Chrome | master on Chrome | VS Code |
|---|---|---|
| ![details-dismiss-0 8](https://user-images.githubusercontent.com/5047891/27468780-2a2d86e0-57ec-11e7-9195-b1c8cc97a6ab.gif) | ![details-dismiss-master](https://user-images.githubusercontent.com/5047891/27468782-30879f08-57ec-11e7-8f93-7e7bf6fe5c2f.gif) | ![details-dismiss-vscode](https://user-images.githubusercontent.com/5047891/27468785-35554170-57ec-11e7-8151-e928911f0bd4.gif) |




### 3. The details are not positioned under the suggestions if the page is narrow
* test page `test/mouse-scrollable-body.html` with an edit to change the language from `xml` to `javascript`.

| v0.8 on Chrome | master on Chrome | VS Code |
|---|---|---|
| ![image](https://user-images.githubusercontent.com/5047891/27468580-d7cd3f40-57ea-11e7-9fb0-b23ba3cec3b7.png) | ![image](https://user-images.githubusercontent.com/5047891/27468484-642413e8-57ea-11e7-9345-888cc12eef9d.png) | ![image](https://user-images.githubusercontent.com/5047891/27468375-be203b8e-57e9-11e7-8447-48bc19b13a26.png) |

**Comment by ramya-rao-a** [comment]
First issue is fixed with https://github.com/Microsoft/vscode/commit/97331424b12948ddac23e00302e72c8a6fac4591

Second issue, your gif doesn't work, so am not sure what the problem is :(

The third one I can repro only after re-sizing and no refresh. Can you confirm? If that is so, then this is due to https://github.com/Microsoft/vscode/issues/27676. Is this a blocker?

**Comment by ramya-rao-a** [comment]
Turns out fixing https://github.com/Microsoft/vscode/issues/27676 (I just did) won't fix the third issue.
Because changing the window size doesn't mean changing the editor size which is what I was listening to in the fix for https://github.com/Microsoft/vscode/issues/27676.

Any other event you know that I can listen to?

**Comment by jrieken** [comment]
Attaching recording of the recording playing for me 
![jun-28-2017 09-06-13](https://user-images.githubusercontent.com/1794099/27624454-18a1ae94-5be1-11e7-8f17-81d6c74a51df.gif)

**Comment by alexandrudima** [comment]
Re 3:

I typically avoid caching window sizes or dom node positions between paints if there is a likelihood that they have changed. 

So I think it is totally ok if the widget is shown, and the browser is resized while the widget is shown, and then the widget does not *immediately* react to the new browser size (i.e. I don't think there is a need for a listener). But, if the widget is hidden and then shown again, it should adopt to the new window size.

The problem is that very many things can lead to a change, i.e. there could have been a user scroll on some parent dom node, or perhaps the integrator inserts new dom nodes to the page, etc. so IMHO there is no event that can capture all these cases except perhaps some specialized mutation observers, but I don't think we should go down that path.

**Comment by ramya-rao-a** [comment]
> But, if the widget is hidden and then shown again, it should adopt to the new window size.

This is tough to do as well. The widget has no knowledge about the window size. It only has knowledge about the editor size. 

In VS Code, I overcame this problem by re-using the max-width set on the widget by the `editor.layoutContentWidget()`. But in the case of `test/mouse-scrollable-element.html`, the max-width on the widget doesn't react to the window size.

**Comment by rebornix** [comment]
@ramya-rao-a with document's scrollLeft, we should be able to know where the widget is position in current window. And then with `document.documentElement.clientWidth`, we can know the window size and check if there is enough space for rendering the suggestion details side by side.

**Comment by ramya-rao-a** [comment]
@rebornix That would mean accessing the document in the window. `suggestWidget` is not aware of that

**Comment by alexandrudima** [comment]
:+1: 

Let's postpone the corner cases for now. 

The suggest widget works a lot better now with the two changes from @ramya-rao-a . Thank you!

**Comment by belohlavek** [comment]
Is there a way I can force the suggestion to always appear below the cursor? I have some cases where it appears on top of it and it gets cut off by the container overflow

**Comment by ramya-rao-a** [comment]
@belohlavek The suggestions should always appear below the cursor as long as there is space for it.

**Comment by belohlavek** [comment]
@ramya-rao-a is the space based on the wrapper container? I'm creating one/four lines expression editors for one of our products. Kind of excel like. We need to guarantee that hover suggestions appear under the cursor. Is there any way I can force this?

**Comment by ramya-rao-a** [comment]
The decision to show the widget either above or below the cursor is decided by the layout. @alexandrudima can answer if there is a way to force it

**Comment by belohlavek** [comment]
Hey @alexandrudima is there actually a way to achieve this?

**Comment by alexandrudima** [comment]
Closing as this will not be fixed.
